pkg/kubelet/pod_workers.go

```diff
@@ -734,10 +748,71 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
 }
 
+// allowPodStart tries to start the pod and returns true if allowed, otherwise
+// it requeues and returns false.
+func (p *podWorkers) allowPodStart(pod *v1.Pod) bool {
+	if !kubetypes.IsStaticPod(pod) {
+		return true
+	}
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+	if !p.isStaticPodAllowedToStart(pod) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		p.podSyncStatuses[pod.UID].working = false
+		return false
+	}
+	p.allowNextStaticPod(kubecontainer.GetPodFullName(pod))
+	return true
+}
+
+// isStaticPodAllowedToStart returns true if there is no other started static
+// pod with the same full name and pod is the oldest static pod waiting to
+// start.
+func (p *podWorkers) isStaticPodAllowedToStart(pod *v1.Pod) bool {
+	fullname := kubecontainer.GetPodFullName(pod)
+	uid, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		if uid == pod.UID {
+			// pod is already started
+			return true
+		}
+		return false
+	}
+
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return false
+	}
+
+	return pod.UID == waitingToStartPods[0]
+}
+
+// allowNextStaticPod allows next static pod to start for the given fullname
+func (p *podWorkers) allowNextStaticPod(fullname string) {
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingToStartPods[0]
+	p.waitingToStartStaticPodsByFullname[fullname] = waitingToStartPods[1:]
```

Comment:
I *think* we need to iterate through waitingToStartPods and remove any uid that is not in podSyncStatus, then take the one in the first position.

Also, can the current UID be started AND in be in the first waitingToStartPods position?  If so, we may also need to exclude that.

I'd like to see unit tests for allowPodStart that capture various uses cases and describe which aren't solved (in this PR).

Reply:
(full branch coverage on these methods added for a `Test_allowPodStart`)

Reply:
> I'd like to see unit tests for allowPodStart that capture various uses cases and describe which aren't solved (in this PR).

done



> I _think_ we need to iterate through waitingToStartPods and remove any uid that is not in podSyncStatus, then take the one in the first position.
> 
> Also, can the current UID be started AND in be in the first waitingToStartPods position? If so, we may also need to exclude that during iteration.

They may be possible. And I think that it is not desirable to leave podWorkers in an inconsistent state(partially completed pod) at any time except inside of the podLock.

So, fixed podWorkers to be able to complete pod clearly.